server/src/main/java/org/opensearch/action/search/SearchPhaseExecutionException.java

```diff
@@ -60,11 +60,36 @@ public SearchPhaseExecutionException(String phaseName, String msg, ShardSearchFa
     }
 
     public SearchPhaseExecutionException(String phaseName, String msg, Throwable cause, ShardSearchFailure[] shardFailures) {
-        super(msg, deduplicateCause(cause, shardFailures));
+        super(msg, getEffectiveCause(cause, shardFailures));
         this.phaseName = phaseName;
         this.shardFailures = shardFailures;
     }
 
+    /**
+     * Determines the effective cause for this exception. If an explicit cause is provided and not
+     * duplicated in shard failures, use it. Otherwise, use the first shard failure's cause.
+     */
+    private static Throwable getEffectiveCause(Throwable cause, ShardSearchFailure[] shardFailures) {
+        if (shardFailures == null) {
+            throw new IllegalArgumentException("shardSearchFailures must not be null");
+        }
+        // If explicit cause provided and not duplicated in shard failures, use it
+        if (cause != null) {
+            for (ShardSearchFailure failure : shardFailures) {
+                if (failure.getCause() == cause) {
+                    // Cause is duplicated, but still return it to properly wire the chain
+                    return cause;
+                }
+            }
+            return cause;
+        }
+        // No explicit cause - use first shard failure's cause if available
+        if (shardFailures.length > 0 && shardFailures[0].getCause() != null) {
```

Comment:
Should we do the length check up on line 73 ("... null or empty") and throw an exception there?